### DISCUSSION TITLE
Load/import functions/constant from a file instead  of redefining them everytime.; Edit: Changed to QnA

### DISCUSSION THREAD

**Discussion by BilakshanP** [lcn2/calc#69] [Q&A] [answered]
There could be some functions/constant you probably add again and again in calc whenever you use it. So, it would be better if you could just import them from a file...

Like, in my case i have to define some functions again and again which are very useful to me, examples:

- degree trignometric functions and their arc counterparts
- physical constnats such as epsilon naught, big G etc

so it wold be really cool if you we could have a file like <filename.calc> and a basic import function like:

import(<filename.calc>)

and it would just load/import all the constants and function from that file...

the structure of file would be exactly how u think it should be:

```
define example(a, b){
    return a + b
}

some_var = 123
```

there could be field for pub/priv too ig,, (idk i am poor at designing/explaining)

but u prolly got the point

**Comment by lcn2** [comment, marked as answer]
Thank you for suggestion, @BilakshanP, before we considered in more detail, we want to be sure you're aware of existing capabilities within calc.

There is a calc startup file that is automatically processed when calc starts (unless you give a `-q` to the calc command line).  See the calc man page for details.

The calc startup file is controlled by the **$CALCRC** environment variable, which it is does exist will mean that the file `.calcrc` will be searched for along a path. People usually put a `.calcrc` file in their home directory for a general purpose, although you may have a `.calcrc` file in your local directory that will take priority.

Let's assume you have a `~/.calcrc` file (in your home directory). If you put those commands you mentioned above into that file, calc will automatically try to execute them on startup (unless you run calc with `-q`). 

For example, you could set change display and default precision when calc start by adding these to your `~/.calcrc` file:

```
display(50);
epsilon(1e-50);
```

Now you may not want those commands to spew out details of every time calc starts. There are ways to silence them.  For example instead of saying adding to your calcrc file:

You could instead add:

```
display(50),;
epsilon(1e-50),;
```

Similarly, when you define functions, you can set a configuration state to not print the fact that the functions to find and then unset it afterwards. I know some people prefer to know that functions are being define when calc is starting up and that's OK it'll do that automatically. I'm just pointing out you can disable that in various ways if you wish.

Now, if you want to have some calc script or situation load commands when you type something, you could use the read command. 

Say you have those commands you listed above in the file "foo.cal". By entering into calc:

```
read "foo.cal"
```

They will be "imported" as you say.

You can even put that command into another file, that when executed will read ("import") those commands as well. 

Some people like to use the `-once` option of read. This will ignore the file if it's been read in your current calc session already.  So:

```
read -once "foo.cal"
```

maybe be what you want to do. 

You mentioned that you wanted:

```
- physical constnats such as epsilon naught, big G etc
```

You could create a "physics.cal" file containing various constants and load them into your calc session by:

```
; read "physics.cal"
```

Again, you might want to silence variable settings by using in "physics.cal" commands of the form such as:

```
G=6.67430e-11;
c=299792458;
e0=8.854187817e-12;
alpha=0.00729735256286;
```

You could even add to your calc startup file:

```
read -once "/some/path/physics.cal"
```

to always have them.

You mentioned that you wanted to do:

```
- degree trignometric functions and their arc counterparts
```

Did you know that calc now has built in functions such as `d2r`?  For example:

```
; print sin(d2r(30)), cos(d2r(60)), tan(d2r(45))
    0.5 0.5 1
```

See the help command in calc for details:

```
; help d2r
```

For a complete list of calc buildin functions:

```
; help builtin
```

For more information on things like the environment variable **$CALCRC**, see:

```
; help environment
```

For help on the above mentioned read command (and other useful commands) see:

```
; help command
```

For help with help, including some ideas of other things you can get help on, try:

```
; help help
```

See also the calc man page. 

You could even define your own sine in degrees function such as:

```
define dsin(x) = sin(d2r(x));
```

And, of course, you can put that in your calc start up file `~/.calcrc`to have such functions automatically available to you.

We hope this helps. If there are additional request or concerns or something that you didn't quite understand or something else related to your original request, please follow up with a comment.

**Reply by BilakshanP** [reply]
Thank you, thank you, thank you!!!

I was searching "Import"/"Load" keywords in the man page or on internet maybe that's why i didn't found what i was looking for!!

You cleared all my doubts, and yes that's exactly what I meant!

Again thanks for your time and this precious brief explanation, and a bit of apologies.

<3 Calc is super, thanks for creating it...

**Reply by lcn2** [reply]
> Thank you, thank you, thank you!!!
> 
> I was searching "Import"/"Load" keywords in the man page or on internet maybe that's why i didn't found what i was looking for!!
> 
> You cleared all my doubts, and yes that's exactly what I meant!
> 
> Again thanks for your time and this precious brief explanation, and a bit of apologies.
> 
> <3 Calc is super, thanks for creating it...

You are most welcome 🙏, best wishes @BilakshanP

**Reply by BilakshanP** [reply]
Umm, so I decided to upgrade calc to lastest version to try d2r and all the other wonders u mentioned but after using `apt` on ubuntu and I found out that the latest version there is `2.12.7.2`.

Then, I went for building it myself from source, the build was successful but now history is not working, any Idea?

Any error: https://pastebin.com/HKJVwh7h
Build logs: https://pastebin.com/ugQEguaT (1 week validity)

**Reply by lcn2** [reply]
> Umm, so I decided to upgrade calc to lastest version to try d2r and all the other wonders u mentioned but after using `apt` on ubuntu and I found out that the latest version there is `2.12.7.2`.
> 
> 
> 
> Then, I went for building it myself from source, the build was successful but now history is not working, any Idea?
> 
> 
> 
> Any error: https://pastebin.com/HKJVwh7h

There's nothing serious with the above warnings. They are a known warnings (that are not known to cause problems and are also unrelated to history) from an hyperactive compiler. 

> Build logs: https://pastebin.com/ugQEguaT (1 week validity)

The build log looks normal too.  

Our offhand guess is that in order to compile calc with history, you need to have _readline_ related packages (from what we can see, this also brings in libhistory on Debian) and the and _ncurses_ related apt packages installed. Once they are installed you may have to recompile calc:

```sh
make clobber all chk
```

Be sure that the READLINE facility is enabled in the top level Makefile:

```make
...
USE_READLINE= -DUSE_READLINE
#
#READLINE_LIB=
#READLINE_EXTRAS=
#
READLINE_LIB= -lreadline
READLINE_EXTRAS= -lhistory -lncurses
...
```

**Reply by BilakshanP** [reply]
Yes, I was getting an dependency unmet error before so I did install(?) `libreadline-dev` for it, `libncurses-dev` was already fulfilled and also `READLINE` facility is enabled in the Makefile.

I'll try to fix it *somehow* if it doesn't work I'll let you know.
